### PR TITLE
Quantum Subroutine

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** Machine generated stablehlo is known to generate functions in the 10k LOC. MLIR compilation strategy revolves around iterating through instructions. Since MLIR operations are a linked list, iterating through this list will trash the cache. A potential solution we found a long time ago was to reduce the number of instructions. This can be achieved through the user denoting the use of control flow but there are still many places where a function would make sense.

**Description of the Change:** This PR introduces the notion of quantum subroutines. In JAXPR, quantum subroutines are analogous to `pjit_p` but just with a different primitive name. This will allow the generation of `func.func` ops during their lowering. This feature is currently only available through PLxPR. It works as follows:

1. generate a wrapper around functions which will call `jax.jit` on a given function.
```python

@subroutine
def foo(): ...
```

2. PLxPR does not have quantum registers at all, so during the conversion of PLxPR to JAXPR with quantum primitives we will insert a register as a parameter and inline the contents of the subroutine.

```python
def flat_foo(*qreg_args):
   qreg, *args = *qreg_args
   retval = interpreter.evaluate(qreg, args)
   return qreg, *retvals
```

3. Once this gets converted to JAXPR with quantum primitives, then it is up to Catalyst to lower it to machine code.

**Benefits:** Faster compilation if used, but may limit the runtime speed in case inlining could optimize the function. I think this is a good trade off since we know compilation is a bottle neck for machine generated code.

**Possible Drawbacks:** Only usable from PLxPR.

**Related GitHub Issues:**

- [ ] A decorator to get the MLIR of only quantum subroutines would be useful for program capture / transformations.
- [ ] Pipeline from PLxPR to Catalyst JAXPR is missing gradient and callback support, so that is also missing from subroutines
- [ ] Subroutines + Gradients is untested and it is not expected to work.

[sc-85745]

### CODE REVIEW

File: frontend/catalyst/from_plxpr.py [hunk 1]
```diff
@@ -301,12 +303,13 @@ class PLxPRToQuantumJaxprInterpreter(PlxprInterpreter):
...
     """
 
-    def __init__(self, device, shots, qreg):
+    def __init__(self, device, shots, qreg, cache):
         self.device = device
         self.shots = shots
         self.wire_map = {}
         self.qreg = qreg
         self.actualized = True
+        self.subroutine_cache = {}
```
Comment (paul0403) [inline frontend/catalyst/from_plxpr.py]:
```suggestion
        self.subroutine_cache = cache
```

File: frontend/test/lit/test_quantum_subroutine.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+    @subroutine
+    def add_one(x):
+        return x + 1
+
+    qml.capture.enable()
+
+    @qml.qjit
+    def main():
+        # CHECK: %{{.*}} = call @add_one(%{{.*}}) : (tensor<i64>) -> tensor<i64>
```
Comment (paul0403) [inline frontend/test/lit/test_quantum_subroutine.py:42]: Can you add the `call` lines to all the other tests? Since this is the point of the feature.

File: frontend/catalyst/from_plxpr/from_plxpr.py [hunk 3]
```diff
@@ -399,6 +404,67 @@ def __call__(self, jaxpr, *args):
...
+        converted_closed_jaxpr_branch = ClosedJaxpr(
+            convert_constvars_jaxpr(converted_jaxpr_branch), ()
+        )
+        self.subroutine_cache[plxpr] = converted_closed_jaxpr_branch
+    else:
+        converted_closed_jaxpr_branch = transformed
+    # jaxpr = from_plxpr(jaxpr)(AbstractQreg(), *args)
+    # So, what I need to do here is transform this jaxpr
+    # With `from_plxpr` to but we need to make sure that
+    # the first argument is treated as the qreg...
```
Comment (erick-xanadu) [inline frontend/catalyst/from_plxpr/from_plxpr.py]:
```suggestion
```

Comment (paul0403) [reply frontend/catalyst/from_plxpr/from_plxpr.py]: Let me know if you need help with the new qreg manager : )

Comment (erick-xanadu) [reply frontend/catalyst/from_plxpr/from_plxpr.py]: Oh I think it is already done! This comment is just something that I still had, but I need to delete. Thank you!

File: frontend/catalyst/from_plxpr/from_plxpr.py [hunk 4]
```diff
@@ -399,6 +404,63 @@ def __call__(self, jaxpr, *args):
...
+        inline=kwargs["inline"],
+        compiler_options_kvs=kwargs["compiler_options_kvs"],
+    )
+
+    self.qreg = vals_out[0]
+    self.qreg_manager = QregManager(self.qreg)
+    vals_out = vals_out[1:]
+
+    for orig_wire, _ in backup:
+        self.qreg_manager[orig_wire] = qextract_p.bind(self.qreg_manager.get(), orig_wire)
```
Comment (paul0403) [inline frontend/catalyst/from_plxpr/from_plxpr.py]: Oh, I have an `extract` method on the manager itself that adds to the map automatically!

Comment (erick-xanadu) [reply frontend/catalyst/from_plxpr/from_plxpr.py]: Please add the suggestion here :)

Comment (paul0403) [reply frontend/catalyst/from_plxpr/from_plxpr.py]: (Instead of going through `__setitem__` here manually)

Comment (paul0403) [reply frontend/catalyst/from_plxpr/from_plxpr.py]:
```suggestion
        self.qreg_manager.extract(orig_wire)
```

File: frontend/catalyst/from_plxpr/from_plxpr.py [hunk 5]
```diff
@@ -55,8 +57,10 @@
     probs_p,
     qalloc_p,
     qdealloc_p,
+    qextract_p,
```
Comment (erick-xanadu) [inline frontend/catalyst/from_plxpr/from_plxpr.py]:
```suggestion
```

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -293,6 +293,9 @@
   This runtime stub is currently for mock execution only and should be treated as a placeholder
   operation. Internally, it functions just as a computational-basis measurement instruction.
 
+* Quantum subroutine stub.
```
Comment (dime10) [inline doc/releases/changelog-dev.md]: Comment as a reminder

File: frontend/catalyst/jax_primitives.py [hunk 7]
```diff
@@ -306,6 +311,55 @@ class MeasurementPlane(Enum):
...
+
+    The problem with this approach is that it will also change the definitions of jax.jit
+    which may be purely classical. The best solutions would be to have two primitives.
+
+    But in order to have two primitives, we would need to replace pjit_p selectively,
+    depending on whether or not we are inside of a qml.qjit context. This would involve
+    rewriting some methods inside JAX or copying them and modifying that one key aspect.
+
+    I, @erick-xanadu, am currently unsure about what is the best way to do this.
+    Maybe this is sufficiently good?
```
Comment (dime10) [inline frontend/catalyst/jax_primitives.py]: I don't know why but this made me laugh 😆 It sounds like you're about to make a solemn declaration

File: frontend/catalyst/jax_primitives.py [hunk 8]
```diff
@@ -306,6 +311,55 @@ class MeasurementPlane(Enum):
...
+            (
+                jax._src.pjit,
+                "pjit_p",
+                old_pjit,
+            ),
+        ):
+            return func(*args, **kwargs)
+
+    @functools.wraps(inside)
+    def wrapper(*args, **kwargs):
```
Comment (dime10) [inline frontend/catalyst/jax_primitives.py:370]: Wow this double wrapper is confusing 😅 So your approach patches pjit to the quantum subroutine, and then inside the decorated function back to jax's pjit? 

Could you elaborate why the patching is required at all? If the subroutine primitive is already identical to pjit in every way but name, can't we use it directly? Or are you saying that is the only way to get it not to hoist constants, even though they are identical except in name? 🤔

File: frontend/catalyst/jax_primitives.py [hunk 9]
```diff
@@ -2333,6 +2387,15 @@ def _cos_lowering2(ctx, x, accuracy):
     return _nary_lower_hlo(hlo.cosine, ctx, x, accuracy=accuracy)
 
 
+def subroutine_lowering(*args, **kwargs):
+    """This is just a method that forwards arguments to _pjit_lowering
+
+    Useful for testing.
```
Comment (dime10) [inline frontend/catalyst/jax_primitives.py]: This says it's for testing, but isn't this the lowering you actually registered? Then it's used in production right?

File: frontend/catalyst/from_plxpr/from_plxpr.py [hunk 10]
```diff
@@ -399,6 +403,64 @@ def __call__(self, jaxpr, *args):
         return self.eval(jaxpr.jaxpr, jaxpr.consts, *args)
 
 
+@PLxPRToQuantumJaxprInterpreter.register_primitive(quantum_subroutine_p)
+def handle_subroutine(self, *args, **kwargs):
+    """
+    Transform the subroutine from PLxPR into JAXPR with quantum primitives"""
```
Comment (dime10) [inline frontend/catalyst/from_plxpr/from_plxpr.py]:
```suggestion
    """
    Transform the subroutine from PLxPR into JAXPR with quantum primitives.
    """
```

File: frontend/catalyst/from_plxpr/from_plxpr.py [hunk 11]
```diff
@@ -399,6 +403,64 @@ def __call__(self, jaxpr, *args):
...
+        converted_closed_jaxpr_branch = ClosedJaxpr(
+            convert_constvars_jaxpr(converted_jaxpr_branch), ()
+        )
+        self.subroutine_cache[plxpr] = converted_closed_jaxpr_branch
+    else:
+        converted_closed_jaxpr_branch = transformed
+
+    # quantum_subroutine_p.bind
+    # is just pjit_p with a different name.
+    vals_out = quantum_subroutine_p.bind(
```
Comment (dime10) [inline frontend/catalyst/from_plxpr/from_plxpr.py:439]: This method is registered against `quantum_subroutine_p`, so we are just mapping a `quantum_subroutine_p` to a slightly different `quantum_subroutine_p` where the body has been converted, right? (was just surprised seeing a primitive mapped to itself)

File: frontend/catalyst/from_plxpr/from_plxpr.py [hunk 12]
```diff
@@ -399,6 +403,64 @@ def __call__(self, jaxpr, *args):
...
+        donated_invars=kwargs["donated_invars"],
+        ctx_mesh=kwargs["ctx_mesh"],
+        name=kwargs["name"],
+        keep_unused=kwargs["keep_unused"],
+        inline=kwargs["inline"],
+        compiler_options_kvs=kwargs["compiler_options_kvs"],
+    )
+
+    self.qreg = vals_out[0]
+    self.qreg_manager = QregManager(self.qreg)
```
Comment (dime10) [inline frontend/catalyst/from_plxpr/from_plxpr.py]: Is the intended usage pattern to instantiate new managers whenever a new qreg value is produced, or is there a way to update it in the existing manager instead?

**Review by dime10 — APPROVED** [review]
Neat, very clean solution 👍

File: frontend/catalyst/from_plxpr/from_plxpr.py (hunk 12 above)
Comment (paul0403) [reply frontend/catalyst/from_plxpr/from_plxpr.py]: > Is the intended usage pattern to instantiate new managers whenever a new qreg value is produced

The qreg manager docstring explains the intended usage. Briefly, 
- For qreg values at the root of a def-use chain (so allocs and block arguments), a new instance needs to be made
- For regular ops that takes in qregs and produces qregs (control flow, quantum adjoint, and I guess subroutines), this would be an update on an existing instance.

File: frontend/catalyst/jax_primitives.py (hunk 8 above)
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py:370]: > So your approach patches pjit to the quantum subroutine, and then inside the decorated function back to jax's pjit?

Correct.

> Could you elaborate why the patching is required at all?

This is necessary because `jax` could introduce other `jax.jit`s. This would mean that `jax.jit` either by the user or by `jax` itself would appear as `subroutine_p`.

File: frontend/catalyst/from_plxpr/from_plxpr.py (hunk 12 above)
Comment (paul0403) [reply frontend/catalyst/from_plxpr/from_plxpr.py]: (Although I'm not sure whether it's a root SSA def-use chain value here specifically)

File: frontend/catalyst/jax_primitives.py (hunk 9 above)
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py]: I want it here, but I am not opposed to removing it. If there's an error, I would like the stack trace to show where it comes from in Catalyst as soon as possible. Here, at least we know that the lowering comes from Catalyst. Happy to remove it if you insist. (I suppose, it is useful for tracing and having the stack trace show the lowering is here)

File: frontend/catalyst/from_plxpr/from_plxpr.py (hunk 11 above)
Comment (erick-xanadu) [reply frontend/catalyst/from_plxpr/from_plxpr.py:439]: Yes, as we are tracing the program, we are binding to quantum subroutine, but with different `call_jaxpr` and different parameters and return arguments (the quantum register).

File: frontend/catalyst/from_plxpr/from_plxpr.py (hunk 12 above)
Comment (erick-xanadu) [reply frontend/catalyst/from_plxpr/from_plxpr.py]: I'll let @paul0403 add suggestions :)

File: frontend/catalyst/jax_primitives.py (hunk 9 above)
Comment (dime10) [reply frontend/catalyst/jax_primitives.py]: I'm not suggesting removing it, I was just inquiring about the comment. I agree having our own function here is useful, since we defined our own primitive!

File: frontend/catalyst/from_plxpr/from_plxpr.py (hunk 12 above)
Comment (dime10) [reply frontend/catalyst/from_plxpr/from_plxpr.py]: Thanks Paul, sounds like an update would be more appropriate here then if I understand correctly?

File: frontend/catalyst/jax_primitives.py (hunk 7 above)
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py]: Technical debt is a serious topic :military_helmet:

File: frontend/catalyst/jax_primitives.py (hunk 8 above)
Comment (dime10) [reply frontend/catalyst/jax_primitives.py:370]: Right, we definitely wouldn't want to perma patch `pjit_p` to `quantum_subroutine_p` (even locally within a qjit scope), but is there a way that doesn't require patching it all? 🤔

Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py:370]: To elaborate, we don't want to treat pjit_p the same was as subroutine_p. If we did, we would try to add the quantum register to purely classical pjit_p-ed functions.

Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py:370]: > Right, we definitely wouldn't want to perma patch pjit_p to quantum_subroutine_p (even locally within a qjit scope), but is there a way that doesn't require patching it all?

The only other way I saw was copying a substantial amount of the callgraph starting from `jax.jit`.

File: frontend/catalyst/from_plxpr/from_plxpr.py (hunk 12 above)
Comment (erick-xanadu) [reply frontend/catalyst/from_plxpr/from_plxpr.py]: Looks like it, what API would be used for an update?

File: frontend/catalyst/jax_primitives.py (hunk 8 above)
Comment (erick-xanadu) [reply frontend/catalyst/jax_primitives.py:370]: Essentially mini-forking jax.

File: frontend/catalyst/from_plxpr/from_plxpr.py (hunk 12 above)
Comment (paul0403) [reply frontend/catalyst/from_plxpr/from_plxpr.py]:
```suggestion
    self.qreg_manager.set(self.qreg)
```

File: frontend/catalyst/from_plxpr/from_plxpr.py [hunk 13]
```diff
@@ -399,6 +403,65 @@ def __call__(self, jaxpr, *args):
...
```
Comment (erick-xanadu) [inline frontend/catalyst/from_plxpr/from_plxpr.py]:
```suggestion
    self.qreg_manager.set(vals_out[0])
```

@paul0403

File: frontend/catalyst/from_plxpr/from_plxpr.py (hunk 12 above)
Comment (paul0403) [reply frontend/catalyst/from_plxpr/from_plxpr.py]: A quick note that updating the qreg value in the manager requires that the old one has been actualized (by `insert_all_danglin_qubits`). This is because like qubit values, when an op like 
`%new_qreg = op %old_qreg`
appears, the old value is essentially expired, and qubits extracted but not inserted into the old one are now bad.

I also print a nice error message if the actualize was missing : )

File: frontend/catalyst/jax_primitives.py (hunk 9 above)
Comment (dime10) [reply frontend/catalyst/jax_primitives.py]:
```suggestion
    Even though we could register the `pjit_p` lowering directly, this makes the code origin
    apparent in stack traces and similar use cases.
```

File: frontend/catalyst/jax_primitives.py (hunk 8 above)
Comment (dime10) [reply frontend/catalyst/jax_primitives.py:370]: Oh got it. So you wanted to invoke a bunch the code from `jax.jit`, because just calling `quantum_subroutine_p.bind()` is not sufficient, that makes sense! (not sure why I only thought about the primitive and not the frontend function part)

File: doc/releases/changelog-dev.md (hunk 6 above)
Comment (erick-xanadu) [inline doc/releases/changelog-dev.md]:
```suggestion
* Support for quantum subroutines was added.
  This feature is expected to improve compilation times for large quantum programs.
```

Comment (erick-xanadu) [reply doc/releases/changelog-dev.md]: I added it here: https://github.com/PennyLaneAI/catalyst/pull/1806/commits/5ac538003612597294505129fc2f333cab6664a5

Comment (erick-xanadu) [reply doc/releases/changelog-dev.md]: I changed the docstring today. Please review and resolve this conversation. I'll merge later today. :+1:

Comment (erick-xanadu) [reply doc/releases/changelog-dev.md]: I'm just going to resolve and merge :)

Comment (dime10) [reply doc/releases/changelog-dev.md]: Sorry, just had a look but I think it will need more info showing the user how to use it :)

Comment (josh146) [reply doc/releases/changelog-dev.md]: Yep I would agree here, the entry could benefit from examples and any nuance that is useful for a user to know! E.g., there are a few things in the PR comment that would be beneficial to have in the changelog, such as 

> Note that using subroutines will result in faster compilation if used, but as a consequence may limit the runtime speed due to/in cases of etc.

Comment (josh146) [reply doc/releases/changelog-dev.md]: Actually the docstring does a great job of this, could basically just copy and paste the docstring
